src/service/SSHKeyForwardingService.ts

```diff
@@ -0,0 +1,216 @@
+      console.log(`[SSH Forwarding] Executing approved push ${pushId}`);
+
+      // Get push details from database
+      const push = await getPush(pushId);
+      if (!push) {
+        console.error(`[SSH Forwarding] Push ${pushId} not found`);
+        return false;
+      }
+
+      if (!push.authorised) {
+        console.error(`[SSH Forwarding] Push ${pushId} is not authorised`);
+        return false;
+      }
+
+      // Check if we have SSH key information
+      if (push.protocol !== 'ssh') {
+        console.log(`[SSH Forwarding] Push ${pushId} is not SSH, skipping key forwarding`);
+        return await this.executeHTTPSPush(push);
+      }
+
+      // Try to get the SSH key from the agent
+      let privateKey = this.sshAgent.getPrivateKey(pushId);
+      let decryptedBuffer: Buffer | null = null;
+
+      if (!privateKey && push.encryptedSSHKey && push.sshKeyExpiry) {
+        const expiry = new Date(push.sshKeyExpiry);
+        const decrypted = SSHKeyManager.decryptSSHKey(push.encryptedSSHKey, expiry);
+        if (decrypted) {
+          console.log(
+            `[SSH Forwarding] Retrieved encrypted SSH key for push ${pushId} from storage`,
+          );
+          privateKey = decrypted;
+          decryptedBuffer = decrypted;
+        }
+      }
+
+      if (!privateKey) {
+        console.warn(
+          `[SSH Forwarding] No SSH key available for push ${pushId}, falling back to proxy key`,
+        );
+        return await this.executeSSHPushWithProxyKey(push);
+      }
+
+      try {
+        // Execute the push with the user's SSH key
+        return await this.executeSSHPushWithUserKey(push, privateKey);
+      } finally {
+        if (decryptedBuffer) {
+          decryptedBuffer.fill(0);
+        }
+        this.removeSSHKeyForPush(pushId);
+      }
+    } catch (error) {
+      console.error(`[SSH Forwarding] Failed to execute approved push ${pushId}:`, error);
+      return false;
+    }
+  }
+
+  /**
+   * Execute SSH push using the user's private key
+   * @param {any} push The push object
+   * @param {Buffer} privateKey The user's SSH private key
+   * @return {Promise<boolean>} True if successful
+   */
+  private static async executeSSHPushWithUserKey(push: any, privateKey: Buffer): Promise<boolean> {
+    try {
+      // Create a temporary SSH key file
+      const tempDir = await fs.promises.mkdtemp(path.join(os.tmpdir(), 'git-proxy-ssh-'));
+      const keyPath = path.join(tempDir, 'id_rsa');
+
+      try {
+        // Write the private key to a temporary file
+        await fs.promises.writeFile(keyPath, privateKey, { mode: 0o600 });
+
+        // Set up git with the temporary SSH key
+        const originalGitSSH = process.env.GIT_SSH_COMMAND;
+        process.env.GIT_SSH_COMMAND = `ssh -i ${keyPath} -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null`;
+
+        // Execute the git push
+        const gitRepo = simpleGit(push.proxyGitPath);
+        await gitRepo.push('origin', push.branch);
```

Comment:
Should the git operations have a timeout in case of disconnections? Might want to test if that plays nice with larger repos too.